### PR TITLE
Tencent Cloud 201921

### PR DESCRIPTION
**Pull request by Akm0d**
### What does this PR do?
Merge https://github.com/saltstack/salt/pull/54518 into 2019.2.1

### CODE REVIEW

**Review by garethgreenaway — APPROVED** [review]

File: doc/topics/cloud/tencentcloud.rst [hunk 1]
```diff
@@ -0,0 +1,303 @@
+=================================
+Getting Started With TencentCloud
+=================================
+
+Tencent Cloud is a secure, reliable and high-performance cloud compute service
+provided by Tencent. It is the 2st largest Cloud Provider in China.
```
Comment (waynew) [inline doc/topics/cloud/tencentcloud.rst]: 2st?

File: tests/unit/test_config.py [hunk 2]
```diff
@@ -1131,7 +1142,11 @@ def test_apply_cloud_providers_config_extend_multiple(self):
                           {'password': 'new-password',
                            'extends': 'my-dev-envs:linode',
                            'location': 'Salt Lake City'
-                          }],
+                          },
+                          {'extends': 'my-dev-envs:tencentcloud',
+                           'id': 'new-id',
+                           'key': 'new-key',
+                           'location': 'ap-beijing'}],
```
Comment (waynew) [inline tests/unit/test_config.py]: if you use 
```suggestion
                           'location': 'ap-beijing'},
        ],
```
(not sure about indent levels) the next change won't have a silly blame - like `),` that you'll have on line 1145 :trollface:

File: tests/integration/files/conf/cloud.profiles.d/tencentcloud.conf [hunk 3]
```diff
@@ -0,0 +1,8 @@
+tencentcloud-test:
+    provider: tencentcloud-config
+    availability_zone: ap-beijing-1
+    image: img-31tjrtph
+    size: S1.SMALL1
+    allocate_public_ip: true
+    internet_max_bandwidth_out: 1
+    password: "@m5$z%.^8&9*as"
```
Comment (Akm0d) [inline tests/integration/files/conf/cloud.profiles.d/tencentcloud.conf]: If this is a legitimate password it needs changing right away :/

File: tests/unit/test_config.py [hunk 4]
```diff
@@ -1140,7 +1155,11 @@ def test_apply_cloud_providers_config_extend_multiple(self):
                            'driver': 'ec2'},
                           {'apikey': 'abcdefghijklmnopqrstuvwxyz',
                            'password': 'supersecret',
-                           'driver': 'linode'}],
+                           'driver': 'linode'},
+                          {'id': 'the-tencentcloud-id',
+                           'location': 'ap-beijing',
+                           'key': 'the-tencentcloud-key',
+                           'driver': 'tencentcloud'}],
```
Comment (waynew) [inline tests/unit/test_config.py]: and here

File: tests/unit/test_config.py [hunk 5]
```diff
@@ -1168,8 +1193,14 @@ def test_apply_cloud_providers_config_extend_multiple(self):
...
                          'id': 'ABCDEFGHIJKLMNOP',
-                         'location': 'ap-southeast-1'}}}
-        self.assertEqual(ret, sconfig.apply_cloud_providers_config(
+                         'location': 'ap-southeast-1'},
+                    'tencentcloud':
+                        {'id': 'the-tencentcloud-id',
+                         'key': 'the-tencentcloud-key',
+                         'location': 'ap-beijing',
+                         'profiles': {},
+                         'driver': 'tencentcloud'}}}
```
Comment (waynew) [inline tests/unit/test_config.py]: also here...

File: tests/unit/test_config.py [hunk 6]
```diff
@@ -1246,7 +1283,12 @@ def test_apply_cloud_providers_extends_not_in_providers(self):
                            'user': 'user@mycorp.com',
                            'location': 'ap-southeast-1',
                            'key': 'supersecretkeysupersecretkey',
-                           'driver': 'linode'}],
+                           'driver': 'linode'},
+                          {'id': 'a-tencentcloud-id',
+                           'key': 'a-tencentcloud-key',
+                           'location': 'ap-guangzhou',
+                           'driver': 'tencentcloud'
+                          }],
```
Comment (waynew) [inline tests/unit/test_config.py]: and all of these

**Review by waynew — CHANGES_REQUESTED** [review]

File: tests/unit/test_config.py [hunk 7]
```diff
@@ -1124,24 +1134,35 @@ def test_apply_cloud_providers_config_extend_multiple(self):
...
+                  'location': 'ap-southeast-1',
+                  'key': 'supersecretkeysupersecretkey',
+                  'driver': 'ec2'},
+                 {'apikey': 'abcdefghijklmnopqrstuvwxyz',
+                  'password': 'supersecret',
+                  'driver': 'linode'},
+                 {'id': 'the-tencentcloud-id',
+                  'location': 'ap-beijing',
+                  'key': 'the-tencentcloud-key',
+                  'driver': 'tencentcloud'}
```
Comment (waynew) [inline tests/unit/test_config.py]: You should add a trailing comma `},` here

File: tests/unit/test_config.py [hunk 8]
```diff
@@ -1081,18 +1080,23 @@ def test_apply_cloud_providers_config_extend(self):
...
                           {'apikey': 'abcdefghijklmnopqrstuvwxyz',
                            'password': 'supersecret',
                            'driver': 'linode'
-                          }],
+                           },
+                          {'id': 'a-tencentcloud-id',
+                           'key': 'a-tencentcloud-key',
+                           'location': 'ap-guangzhou',
+                           'driver': 'tencentcloud'
+                           }],
```
Comment (waynew) [inline tests/unit/test_config.py]: Did you mean  to leave this, or do you want to make it
```
  },
]
```
(TBH, I'd probably just reformat this whole my-dev-envs to something like...)

```
'my-dev-envs': [
    {'id': 'ABCD...',
     'user': ...
     },
    {'apikey': ...
     },
    {'id': '...',
     },
]
```

File: tests/unit/test_config.py [hunk 9]
```diff
@@ -554,11 +553,8 @@ def test_master_file_roots_glob(self, tempdir, fpath):
             )
         config = sconfig.master_config(fpath)
         base = config['file_roots']['base']
-        self.assertEqual(set(base), set([
-            os.path.join(tempdir, 'a'),
-            os.path.join(tempdir, 'b'),
-            os.path.join(tempdir, 'c')
-        ]))
+        self.assertEqual(set(base),
+                         {os.path.join(tempdir, 'a'), os.path.join(tempdir, 'b'), os.path.join(tempdir, 'c')})
```
Comment (waynew) [inline tests/unit/test_config.py]: Any clue why all of these blocks changed?

I don't *think* I touched any of these :suspect:

File: tests/unit/test_config.py [hunk 10]
```diff
@@ -983,19 +975,29 @@ def test_apply_vm_profiles_config_success(self):
...
+            'conf_file': PATH
+        }
+        ret = {
+            'test-profile': {
+                'profile': 'test-profile',
+                'provider': 'test-provider:digitalocean',
+                'image': 'Ubuntu 12.10 x64',
+                'size': '512MB'
+            }
+        }
```
Comment (waynew) [inline tests/unit/test_config.py]: Same here - I'd probably leave it, but if not, I'd definitely add trailing commas.

File: tests/unit/test_config.py [hunk 11]
```diff
@@ -1032,24 +1057,58 @@ def test_apply_vm_profiles_config_extend_override_success(self):
...
+                'profile': 'dev-instances',
+                'ssh_username': 'test_user',
+                'minion': {
+                    'grains': {
+                        'role': 'webserver'
+                    }
+                },
+                'provider': 'test-config:ec2'
+            }
+        }
```
Comment (waynew) [inline tests/unit/test_config.py]: Bah, missing formats on this one too 😝

**Review by waynew — CHANGES_REQUESTED** [review]

**Comment by Akm0d** [comment]
re-run centos7

**Comment by Akm0d** [comment]
re-run centos7

**Comment by likexian** [comment]
Hello @Akm0d 
Is there anything I can help?

**Comment by Akm0d** [comment]
@likexian we're just waiting for the unrelated test failures to pass or for approval from core

**Comment by Akm0d** [comment]
re-run full amazon2

**Comment by likexian** [comment]
Hello,
The failed CI seems not the responsibility of this PR's code. Can someone please have a review?
It take too long to process this PR.

**Comment by Akm0d** [comment]
```Warning, treated as error:
10:48:08  /var/jenkins/workspace/pr-docs_PR-54526/doc/ref/clouds/all/salt.cloud.clouds.tencentcloud.rst:document isn't included in any toctree```

**Comment by Akm0d** [comment]
re-run full all

**Comment by likexian** [comment]
@Akm0d This PR takes so long, what can I do for it?

**Comment by Akm0d** [comment]
> @Akm0d This PR takes so long, what can I do for it?

I'm sorry this has taken so long :/ We just need all the tests passing.

File: salt/cloud/clouds/tencentcloud.py [hunk 12]
```diff
@@ -0,0 +1,1029 @@
+# -*- coding: utf-8 -*-
+
+'''
+Tencent Cloud Cloud Module
+=============================
+
+.. versionadded:: 2019.09.01
```
Comment (dwoz) [inline salt/cloud/clouds/tencentcloud.py]: 3000?

**Review by dwoz — CHANGES_REQUESTED** [review]

File: salt/cloud/clouds/tencentcloud.py (hunk 12 above)
Comment (Akm0d) [reply salt/cloud/clouds/tencentcloud.py]: :+1:  It looks like the linter might have an issue with 3000 as a version

**Comment by Akm0d** [comment]
re-run full opensuse15

**Comment by Akm0d** [comment]
re-run full amazon1

**Comment by Akm0d** [comment]
re-run full amazon1

**Comment by waynew** [comment]
FWIW, it seems like py2/amazon1 might be/have been pre-deprecated by Amazon.

**Review by waynew — APPROVED** [review]

**Comment by Akm0d** [comment]
re-run full amazon1

**Comment by likexian** [comment]
@Akm0d You are my hero!
It so difficult to make all tests pass. Thanks so much for your works.

**Pull request merged**